File: src/main/java/co/bitshifted/xapps/backstage/deploy/DeploymentProcessTask.java
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.backstage.deploy;

import co.bitshifted.xapps.backstage.BackstageConstants;
import co.bitshifted.xapps.backstage.deploy.builders.DeploymentBuilder;
import co.bitshifted.xapps.backstage.entity.AppDeployment;
import co.bitshifted.xapps.backstage.entity.AppDeploymentStatus;
import co.bitshifted.xapps.backstage.entity.Application;
import co.bitshifted.xapps.backstage.model.CpuArch;
import co.bitshifted.xapps.backstage.model.DeploymentStatus;
import co.bitshifted.xapps.backstage.exception.DeploymentException;
import co.bitshifted.xapps.backstage.model.OS;
import co.bitshifted.xapps.backstage.repository.AppDeploymentRepository;
import co.bitshifted.xapps.backstage.repository.AppDeploymentStatusRepository;
import co.bitshifted.xapps.backstage.repository.ApplicationRepository;
import co.bitshifted.xapps.backstage.util.BackstageFunctions;
import co.bitshifted.xapps.backstage.util.PackageUtil;
import lombok.EqualsAndHashCode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPathExpressionException;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.time.ZonedDateTime;
import java.util.function.Function;

/**
 * @author Vladimir Djurovic
 */
@Slf4j
@EqualsAndHashCode
public class DeploymentProcessTask implements Runnable {

	@EqualsAndHashCode.Include
	private final Path deploymentArchive;
	@EqualsAndHashCode.Include
	private final Path deploymentWorkDir;
	@EqualsAndHashCode.Include
	private String deploymentId;

	@EqualsAndHashCode.Exclude
	@Autowired
	private Function<TargetDeploymentInfo, DeploymentBuilder> deploymentBuilderFactory;

	@EqualsAndHashCode.Exclude
	@Autowired
	private ApplicationRepository applicationRepository;

	@EqualsAndHashCode.Exclude
	@Autowired
	private AppDeploymentRepository appDeploymentRepository;

	public DeploymentProcessTask(File file) {
		this.deploymentArchive = file.toPath();
		this.deploymentWorkDir = deploymentArchive.getParent();
	}

	@Override
	public void run() {
		log.info("Start processing deployment archive {}", deploymentArchive.toFile().getName());
		log.debug("Setting deployment task status to {}", DeploymentStatus.IN_PROGRESS);

		try{
			var deploymentPackageDir = PackageUtil.unpackZipArchive(deploymentArchive);
			// create deployment configuration
			var igniteConfigPath = deploymentPackageDir.resolve(BackstageConstants.IGNITE_CONFIG_FILE_NAME);
			var releaseNumber = BackstageFunctions.getReleaseNumberFromDeploymentDir(deploymentWorkDir.toFile());
			var xmlProcessor = new XmlProcessor(igniteConfigPath, releaseNumber);
			var deploymentConfig = xmlProcessor.getDeploymentConfig();
			// run Mac deployment
			var macDeploymentInfo = TargetDeploymentInfo.builder().deploymentConfig(deploymentConfig)
					.deploymentPackageDir(deploymentPackageDir)
					.targetOs(OS.MAC_OS_X)
					.targetCpuArch(CpuArch.X_64)
					.xmlProcessor(xmlProcessor)
					.build();
			var macDeploymentBuilder = deploymentBuilderFactory.apply(macDeploymentInfo);
			macDeploymentBuilder.createDeployment();
			// run Windows deployment
			var win64DeploymentInfo = TargetDeploymentInfo.builder()
					.deploymentConfig(deploymentConfig)
					.deploymentPackageDir(deploymentPackageDir)
					.targetOs(OS.WINDOWS)
					.targetCpuArch(CpuArch.X_64)
					.xmlProcessor(xmlProcessor)
					.build();
			var win64deploymentBuilder = deploymentBuilderFactory.apply(win64DeploymentInfo);
			win64deploymentBuilder.createDeployment();
			saveDeployment(deploymentConfig.getAppId());
		} catch(Exception ex) {
			log.error("Failed to create deployment package", ex);
			throw new RuntimeException(ex);
		}
	}

	public void init(String deploymentId) {
		this.deploymentId = deploymentId;
		log.info("Initialized deployment task {}", deploymentArchive.getParent().toFile().getName());
	}

	public String getDeploymentId() {
		return deploymentId;
	}

	private void saveDeployment(String applicationId) {
		var app = applicationRepository.findById(applicationId).get();
		var deployment = new AppDeployment();
		deployment.setApplication(app);
		deployment.setReleaseTime(ZonedDateTime.now(BackstageConstants.UTC_ZONE_ID));
		deployment.setReleaseNumber(BackstageFunctions.getReleaseNumberFromDeploymentDir(deploymentWorkDir.toFile()));
		appDeploymentRepository.save(deployment);
	}

}

File: src/main/kotlin/co/bitshifted/backstage/service/deployment/DeploymentProcessTask.kt
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.backstage.service.deployment

import co.bitshifted.backstage.BackstageConstants.DEPLOYMENT_DEPENDENCIES_DIR
import co.bitshifted.backstage.BackstageConstants.DEPLOYMENT_OUTPUT_DIR
import co.bitshifted.backstage.BackstageConstants.DEPLOYMENT_RESOURCES_DIR
import co.bitshifted.backstage.exception.BackstageException
import co.bitshifted.backstage.exception.ErrorInfo
import co.bitshifted.backstage.model.DeploymentStage
import co.bitshifted.ignite.common.model.DeploymentStatus
import co.bitshifted.backstage.model.DeploymentTaskConfig
import co.bitshifted.backstage.repository.DeploymentRepository
import co.bitshifted.backstage.service.ContentService
import co.bitshifted.backstage.service.deployment.builders.DeploymentBuilder
import co.bitshifted.backstage.service.deployment.builders.DeploymentBuilderConfig
import co.bitshifted.backstage.util.Downloader
import co.bitshifted.backstage.util.collectAllDeploymentResources
import co.bitshifted.backstage.util.logger
import co.bitshifted.ignite.common.dto.RequiredResourcesDTO
import co.bitshifted.ignite.common.model.DeploymentStatus.*
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.io.path.exists
import kotlin.io.path.inputStream

const val MAVEN_CENTRAL_REPO_BASE_URL = "https://repo.maven.apache.org/maven2"

class DeploymentProcessTask  (
    val taskConfig: DeploymentTaskConfig
): Runnable {

    private val logger = logger(this)
    private val objectMapper = ObjectMapper()
    @Autowired var contentService : ContentService? = null
    @Autowired var downloader: Downloader? = null
    @Autowired lateinit var deploymentRepository : DeploymentRepository
    @Autowired lateinit var deploymentBuilderFactory : java.util.function.Function<DeploymentBuilderConfig, DeploymentBuilder>

    override fun run() {
        logger.info("Start processing deployment {}", taskConfig.deploymentConfig.deploymentId)
        try {
            if (taskConfig.stage == DeploymentStage.STAGE_ONE) {
                runDeploymentStageOne()
            } else {
                runDeploymentStageTwo()
            }
        } catch(th : Throwable) {
            logger.error("Failed to run deployment process", th)
        }

    }

    private fun runDeploymentStageOne() {
        // download dependencies if not exist
        val requirements = RequiredResourcesDTO()
        taskConfig.deploymentConfig.jvmConfiguration?.collectAllDependencies()?.forEach {
            val exists = contentService?.exists(it.sha256 ?: "unknown", it.size ?: 0) ?: false
            if (!exists) {
                logger.debug("Dependency {} does not exist", it.artifactId)
                val response = downloader?.downloadJavaDependency(MAVEN_CENTRAL_REPO_BASE_URL, it)
                if (response?.second != HttpStatus.OK.value()) {
                    logger.info("Added dependency to missing list")
                    requirements.dependencies.add(it)
                } else {
                    logger.info("Saving dependency {}", it.artifactId)
                    contentService?.save(response?.first)
                }
            }
        }
        // check for missing resources
        val allResources = collectAllDeploymentResources(taskConfig.deploymentConfig)
        allResources.forEach {
            val exists = contentService?.exists(it.sha256 ?: "unknown", it.size ?: 0) ?: false
            if (!exists) {
                logger.info("Adding resource ${it.source} to missing resources list")
                requirements.resources.add(it)
            }
        }

        val deployment = deploymentRepository?.findById(taskConfig.deploymentConfig.deploymentId ?: "unknown")?.orElseThrow { BackstageException(ErrorInfo.DEPLOYMENT_NOT_FOND, taskConfig.deploymentConfig.deploymentId) }
        val text = objectMapper.writeValueAsString(requirements)
        logger.debug("Stage one requirements for deployment ID {}: {}", taskConfig.deploymentConfig.deploymentId, text)
        deployment?.requiredData = text
        deployment?.status = STAGE_ONE_COMPLETED
        deploymentRepository?.save(deployment)

        logger.info("Deployment ID {}, status: {}", taskConfig.deploymentConfig.deploymentId, deployment.status)
    }

    private fun runDeploymentStageTwo() {
        logger.info("Starting deployment stage two...")
        setDeploymentStatus(STAGE_TWO_IN_PROGRESS)
        processFinalContent()
        val outputDir = createDeploymentStructure()
        val builder = deploymentBuilderFactory.apply(DeploymentBuilderConfig( outputDir, taskConfig.deploymentConfig, contentService))
        val success = builder.build()
        setDeploymentStatus( if(success) SUCCESS else FAILED)
    }

    private fun processFinalContent() {
        logger.debug("Checking final dependencies...")
        taskConfig.deploymentConfig.jvmConfiguration?.dependencies?.forEach {
            logger.debug("Checking dependency {}:{}:{}", it.groupId, it.artifactId, it.version)
            val exists = contentService?.exists(it.sha256 ?: "unknown", it.size ?: 0) ?: false
            if (!exists) {
                // copy from deployment directory
                logger.debug("Dependency {}:{}:{} does not exist in storage, trying deployment directory...", it.groupId, it.artifactId, it.version)
                val depPath = Paths.get(taskConfig.contentPath?.toFile()?.absolutePath, DEPLOYMENT_DEPENDENCIES_DIR, it.sha256)
                if(depPath.exists()) {
                    logger.debug("Copying dependency to storage...")
                    contentService?.save(depPath.inputStream())
                    logger.debug("Successfully stored dependency {}:{}:{}", it.groupId, it.artifactId, it.version)
                } else {
                    logger.error("Required dependency {}:{}:{} not found in deployment directory", it.groupId, it.artifactId, it.version)
                }
            }
        }
        val allResources = collectAllDeploymentResources(taskConfig.deploymentConfig)
        val resourceBasePath = Paths.get(taskConfig.contentPath?.toFile()?.absolutePath, DEPLOYMENT_RESOURCES_DIR)
        allResources.forEach {
            logger.debug("Checking resource {}, hash {}", it.target, it.sha256)
            val exists = contentService?.exists(it.sha256 ?: "unknown", it.size ?: 0) ?: false
            if (!exists) {
                logger.debug("Resource {} does not exist in storage, trying deployment directory")
                val resPath = resourceBasePath.resolve(it.target)
                if (resPath.exists()) {
                    logger.debug("Resource path {} exists. Copying resource to storage.", resPath.toFile().absolutePath)
                    contentService?.save(resPath.inputStream())
                    logger.debug("Resource {]  copied to storage", it.target)
                } else {
                    logger.error("Required resource {} not found in deployment directory", it.target)
                }
            }
        }
    }

    private fun createDeploymentStructure() : Path {
        val outputDir = Paths.get(taskConfig.contentPath?.toFile()?.absolutePath, DEPLOYMENT_OUTPUT_DIR)
        Files.createDirectories(outputDir)
        logger.info("Created output directory {}", outputDir.toFile().absolutePath)
        return outputDir
    }

    private fun setDeploymentStatus(status : DeploymentStatus) {
        val deployment = deploymentRepository?.findById(taskConfig.deploymentConfig.deploymentId ?: "unknown") ?. orElseThrow { BackstageException(ErrorInfo.DEPLOYMENT_NOT_FOND, taskConfig.deploymentConfig.deploymentId) }
        deployment.status = status
        deploymentRepository?.save(deployment)
    }
}